src/workerd/api/urlpattern-standard.c++

```diff
@@ -0,0 +1,253 @@
+    } else {
+      result.groups.set(js, k, js.undefined());
+    }
+  }
+  return result;
+}
+
+bool URLPattern::getHasRegExpGroups() const {
+  return inner.has_regexp_groups();
+}
+
+URLPattern::URLPatternResult URLPattern::createURLPatternResult(
+    jsg::Lock& js, const ada::url_pattern_result& other) {
+  URLPatternResult result{
+#define V(_, name) .name = URLPattern::createURLPatternComponentResult(js, other.name),
+    URL_PATTERN_COMPONENTS(V)
+#undef V
+  };
+
+  auto vecInputs = kj::heapArray<kj::OneOf<jsg::JsString, URLPatternInit>>(other.inputs.size());
+  size_t i = 0;
+  for (const auto& input: other.inputs) {
+    if (std::holds_alternative<std::string_view>(input)) {
+      auto raw = std::get<std::string_view>(input);
+      vecInputs[i] = js.str(kj::ArrayPtr(raw.data(), raw.size()));
+    } else {
+      KJ_DASSERT(std::holds_alternative<ada::url_pattern_init>(input));
+      auto obj = std::get<ada::url_pattern_init>(input);
+      vecInputs[i] = createURLPatternInit(obj);
+    }
+    i++;
+  }
+  result.inputs = mv(vecInputs);
+  return result;
+}
+
+#define DEFINE_GETTER(uppercase, lowercase)                                                        \
+  kj::StringPtr URLPattern::get##uppercase() const {                                               \
+    auto value = inner.get_##lowercase();                                                          \
+    return kj::StringPtr(value.data(), value.size());                                              \
+  }
+URL_PATTERN_COMPONENTS(DEFINE_GETTER)
+#undef DEFINE_GETTER
+
+jsg::Ref<URLPattern> URLPattern::constructor(jsg::Lock& js,
+    jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> maybeInput,
+    jsg::Optional<kj::OneOf<kj::String, URLPatternOptions>> maybeBase,
+    jsg::Optional<URLPatternOptions> maybeOptions) {
+  ada::url_pattern_input input;
+  std::optional<std::string_view> base{};
+  std::optional<ada::url_pattern_options> options{};
+
+  KJ_IF_SOME(mi, maybeInput) {
+    KJ_SWITCH_ONEOF(mi) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        input = std::string_view(str.begin(), str.size());
+      }
+      KJ_CASE_ONEOF(init, URLPatternInit) {
+        input = init.toAdaType();
+      }
+    }
+  } else {
+    input = ada::url_pattern_init{};
+  }
+
+  KJ_IF_SOME(b, maybeBase) {
+    KJ_SWITCH_ONEOF(b) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        base = std::string_view(str.begin(), str.size());
+      }
+      KJ_CASE_ONEOF(o, URLPatternOptions) {
+        options = o.toAdaType();
+      }
+    }
+  }
+
+  if (!options.has_value()) {
+    KJ_IF_SOME(o, maybeOptions) {
+      options = o.toAdaType();
+    }
+  }
+
+  std::string_view* base_opt = base ? &(*base) : nullptr;
+  auto result =
+      ada::parse_url_pattern<URLPatternRegexEngine>(input, base_opt, options ? &*options : nullptr);
+  if (result.has_value()) {
+    return jsg::alloc<URLPattern>(std::move(*result));
+  }
+
+  JSG_FAIL_REQUIRE(TypeError, "Failed to construct URLPattern");
```

Comment:
different error handling compared to the other functions, why not be consistent with
`JSG_REQUIRE(result.has_value(), ...);`